float32 distance1
float32 distance2